### PR TITLE
fix issue 15371 - fully debride the traits

### PR DESCRIPTION
**Pull request by ghost**
@MartinNowak has already done it for `getMember` ([after this discussion](https://forum.dlang.org/thread/nq512d$9po$1@digitalmars.com)) but more still remains `getAttributes` or for example `getProtection`, which still triggers a deprecation message when trying to get the visibility of a private symbol. This case is especially nonsensical since the goal can be **exactly** to filter the symbols that are not visible in normal code !

### CODE REVIEW

**Comment by BBasile** [comment]
I've missed the new style to bypass the protection __traits(getProtection, __traits(getMember, stuff)).
The first commit allowed __traits(getProtection, stuff)) but broke the test suite. Couldn't both styles be allowed ?

File: src/ddmd/traits.d [hunk 1]
```diff
@@ -374,12 +374,16 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
         printf("TraitsExp::semantic() %s\n", e.toChars());
     }
 
+    Scope* scx = sc.push();
```
Comment (MartinNowak) [inline src/ddmd/traits.d]: Please use a more obvious name than `scx`.

**Review by MartinNowak — COMMENTED** [review]
We still haven't removed access checks, this should wait until that's done, before we create some intermediate Frankenstein state.

Also not a fan of unspecifically removing visibility for all traits, can you please list which ones should behave like that and why?

It might be too simple to introduce mistakes that way, `e.g. __traits(getFunctionAttributes, Foo.invisibleFunc)`, while `__traits(getFunctionAttributes, __traits(getMember, Foo, "invisibleFunc"))` is clearly distinct from normal dot access.

File: src/ddmd/traits.d (hunk 1 above)
Comment (BBasile) [inline src/ddmd/traits.d]: i've simply moved a block you've added this fall when there was the regression.

> Also not a fan of unspecifically removing visibility for all traits, can you please list which ones should behave like that and why?

The protection is bypassed for everything excepted `__traits(compiles)`. All the other traits are unbridled. The reason for that is that the user decides to respect or not the protection with `__traits(getProtection)`.

**Comment by UplinkCoder** [comment]
this pr sets a bad precedent.
@BBasile isn't there some more systemic way to archive the same ?
Scopes are expensive!

**Comment by BBasile** [comment]
I think so. To change and restore the flag of the scope worked in a first attempt.

**Pull request closed without merging**